README.md

```diff
@@ -1,235 +1,250 @@
+
+Adding a query parameter or controller property fixes this issue.
+
+
+### preserveScrollPosition with queryParams
+
+In certain cases, you might want to have certain routes preserve scroll position when coming from a specific location.
+For example, inside your application, there is a way to get to a route where the user expects scroll position to be
+preserved (such as a tab section).
+
+**1.** Add query param in controller
+
+Add `preserveScrollPosition` as a queryParam in the controller for the route that needs to preserve the scroll position.
+
+Example:
+
+```javascript
+import Controller from '@ember/controller';
+
+export default Controller.extend({
+  queryParams: [
+    'preserveScrollPosition',
+  ],
+});
+```
+
+**2.** Pass in query param
+
+Next, in the place where a transition is triggered, pass in `preserveScrollPosition=true`. For example
+
+```handlebars
+{{link-to "About Tab" 'tab.about' (query-params preserveScrollPosition=true) }}
+```
+
+
+### preserveScrollPosition with a controller property
+
+In other cases, you may have certain routes that always preserve scroll position, or routes where the controller can
+decide when to preserve scroll position. For instance, you may have some nested routes that have true nested UI where
+preserving scroll position is expected. Or you want a particular route to start off with the default scroll-to-top
+behavior but then preserve scroll position when query params change in response to user interaction. Using a controller
+property also allows the use of preserveScrollPosition without adding this to the query params.
+
+
+**1.** Add query param to controller
+
+Add `preserveScrollPosition` as a controller property for the route that needs to preserve the scroll position.
+In this example we have `preserveScrollPosition` initially set to false so that we get our normal scroll-to-top behavior
+when the route loads. Later on, when an action triggers a change to the `filter` query param, we also set
+`preserveScrollPosition` to true so that this user interaction does not trigger the scroll-to-top behavior.
+
+Example:
+
+```javascript
+import Controller from '@ember/controller';
+
+export default Controller.extend({
+  queryParams: ['filter'],
+
+  preserveScrollPosition: false,
+
+  actions: {
+    changeFilter(filter) {
+      this.set('preserveScrollPosition', true);
+      this.set('filter', filter);
+    }
+  }
+});
+```
+
+**2.** Reset preserveScrollPosition if necessary
+
+If your controller is changing the preserveScrollPosition property, you'll probably need to reset
+`preserveScrollPosition` back to the default behavior whenever the controller is reset. This is not necessary on routes
+where `preserveScrollPosition` is always set to true.
+
+```javascript
+import Router from '@ember/routing/route';
+
+export default Route.extend({
+  resetController(controller) {
+    controller.set('preserveScrollPosition', false);
+  }
+});
+```
+
+
+## Running Tests
```

Comment:
This seems like a thing for `CONTRIBUTING.md` (along with the greenkeeper mention)